### ISSUE TITLE
Use TensorShape for shapes in TFE

### ISSUE THREAD

**Issue by yanndupis** [tf-encrypted/tf-encrypted#499]
In the keras code base they use [tf.TensorShape](https://www.tensorflow.org/api_docs/python/tf/TensorShape) at different places, for example [here](https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/keras/layers/core.py#L983). 

We should consider using [tf.TensorShape](https://www.tensorflow.org/api_docs/python/tf/TensorShape) as well, so we can handle properly partially-known shapes and unknown shapes.

**Comment by jvmancuso** [comment]
from @mortendahl in #494 
>one problem with partial shapes is the triple stuff, since they can't be generated offline if the shape is not complete. We could make that a special case though, as the online triple source should work with partial shapes (although we may have to add some control dependencies).

**Comment by jvmancuso** [comment]
yeah, currently the online triples don't work either (factories try to generate tensors with partially known shapes, which gives a ValueError since tensors need to have well-defined shapes)
